### ISSUE TITLE
[google_maps_flutter_web] Ability to animate camera with duration

### ISSUE THREAD

**Issue by jokerttu** [flutter/flutter#159265]
### Use case

User should be able to configure camera update duration on Web platform.

PR https://github.com/flutter/packages/pull/7648 adds support to control camera animation duration to google_maps_flutter_platform_interface and adds implementation for Android and iOS.

Related to #39810

### Proposal

Maps Javascript API does not support setting animation duration for camera controls by default. 
Maps Javascript API -documentation has example how to implement custom animation duration with Tween.js library.
https://developers.google.com/maps/documentation/javascript/examples/move-camera-ease

Similar behaviour could be gained using TickerProvider and stepping camera position between start and end position.

**Comment by brunonahornygaudium** [comment]
@jokerttu 
Was going to implement this by myself for a project 
Thank you! 

If you need any help, let me know

**Comment by ditman** [comment]
Unfortunately, I don't think Google Maps on the web supports any control of the duration of the animation... See docs for moveCamera, panTo and friends:

https://developers.google.com/maps/documentation/javascript/reference/map#Map.moveCamera

**Comment by jokerttu** [comment]
@ditman 
> Unfortunately, I don't think Google Maps on the web supports any control of the duration of the animation... See docs for `moveCamera`, `panTo` and friends:
> 
> https://developers.google.com/maps/documentation/javascript/reference/map#Map.moveCamera

Yes, Google Maps on the web does not provide a direct method to set the duration of an animation. However, animations can be implemented using custom tweening, as suggested in the documentation:
https://developers.google.com/maps/documentation/javascript/examples/move-camera-ease

For Flutter, I have a working solution that uses a ticker to interpolate between the current and target camera positions. However, this approach has certain considerations and was excluded from the related pull request (https://github.com/flutter/packages/pull/7648). I am about to create separate PR later to have separate discussion of the implementation.

One challenge is that the current implementations of `animateCamera` and `moveCamera` both animate the camera, and there is currently no way to move the camera instantly without animation. There are separate methods available for instantly moving the camera, but not for all possible options of the `CameraUpdate` object, and any of these are not currently used/implemented. Also the default behavior of moving/animating camera position detects the move distance and decides whether to move or animate the camera accordingly; which is fine, but differs from other platforms.

If we choose to interpolate the camera position, this could mimic the default behavior. However, it must be clearly documented to warn developers against using interpolation for certain scenarios, such as when the tile loading process cannot keep up with the animation. Same issue is on iOS though.

This feature could be introduced in a separate PR, and there are a couple of approaches to implement it:
1. Fix the moveCamera functionality allowing developers to handle interpolation between camera positions themselves.
2. Support animation duration on platform level and add built-in ticker and interpolation support, having same support as iOS and Android.

**Comment by ditman** [comment]
@jokerttu yyyikesssss! This is demoed with the vector maps, I wonder what happens with the raster tiles :P

> One challenge is that the current implementations of `animateCamera` and `moveCamera` both animate the camera, and there is currently no way to move the camera instantly without animation.

That's probably a legit bug. Once `animateCamera` makes it into the code, and that becomes the "animated" version, we could make `moveCamera` be always immediate. We can maybe roll a major version update or something like that to notify people of the "breaking" behavior.

Maybe all the `newXXX` camera moves should use [`Map.moveCamera`](https://developers.google.com/maps/documentation/javascript/reference/map#Map.moveCamera) instead of `panTo`, [as they do now](https://github.com/flutter/packages/blob/aad08ecb973ea67bd9cd40a88654f3ee9df30781/packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart#L624-L634).

**Comment by jokerttu** [comment]
> [@jokerttu](https://github.com/jokerttu) yyyikesssss! This is demoed only in the vector maps, I wonder what happens with the raster maps :P (It possibly wouldn't work well, I don't think the raster maps support "zoom" levels that are not an int?)^

You can enable fractional zoom levels for raster maps as well using:
https://developers.google.com/maps/documentation/javascript/reference/map#MapOptions.isFractionalZoomEnabled

> > One challenge is that the current implementations of `animateCamera` and `moveCamera` both animate the camera, and there is currently no way to move the camera instantly without animation.
> 
> That's probably a legit bug. Once `animateCamera` makes it into the code, and that becomes the "animated" version, we could make `moveCamera` be always immediate. We can maybe roll a major version update or something like that to notify people of the "breaking" behavior.

Yes, exactly — this would be a breaking change. And since the package is endorsed, it would also introduce a breaking change to the app-facing package, right?

> Maybe all the `newXXX` camera moves should use [`Map.moveCamera`](https://developers.google.com/maps/documentation/javascript/reference/map#Map.moveCamera) instead of `panTo`, [as they do now](https://github.com/flutter/packages/blob/aad08ecb973ea67bd9cd40a88654f3ee9df30781/packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart#L624-L634).

Yep, the current implementation for both animateCamera and moveCamera uses the method:
```dart
    case 'scrollBy':
      map.panBy(json[1]! as num, json[2]! as num);
```
which always animates the camera.
The values are given in pixels.
One issue is that there's currently no way to run this without animating the move. It might be possible to calculate the new location in lat/lng and move directly there, but there are many edge cases where this wouldn't work reliably. For example when map is zoomed out, you cannot calculate or move to latlng which are outside of the world coordinates, but panTo works on these cases as well. 🤯 
So if this breaking change is implemented, it might mean that some of these methods simply won’t be supported on the web platform for `moveCamera`?

Also, the default animation behavior differs depending on the distance. When the source and target points are close, it just scrolls. But when they’re far apart, the default behavior zooms out, then zooms in to the new location.

If custom duration is implemented, the animation would, by default, tween directly between the cameraUpdate configuration—without zooming out (which is actually how it works on iOS as well).
When a duration is provided, developers would need to handle zooming out and then zooming in themselves. It would definitely be nice to offer a helper for this, but it’s certainly not something provided by the underlying Maps SDKs.

**Comment by ditman** [comment]
TIL about `isFractionalZoomEnabled`. I'm not even sure it's even needed?

* https://jsfiddle.net/ditman/gcq4Lwy2/ (I tried to force the RASTER map, but I'm not sure if the configuration of the API Key I'm using (taken from the [SDK docs](https://developers.google.com/maps/documentation/javascript/examples/aerial-simple#try-sample)) is overriding that and I'm looking at a vector map :/

> One issue is that there's currently no way to run this without animating the move.

Yeah, I think this is a bug and probably needs to be fixed. As you said, it'd allow users to implement their custom animations at least. I'm a little bit concerned about the performance of us giving the interpolated solution. I wonder if we could have an extra parameter on the animated camera moves that lets users opt in to this interpolated animation, rather than applying it by default. We can make that boolean parameter `false` by default, so the change is "less breaking".

Of course the breaking change would be the `moveTo` never animating, but I'm not sure if that's a breaking change or correcting a feature that is wrong :)

**Comment by mdebbar** [comment]
@jokerttu is this still on your radar?
